Change target Owner on warhead impact

Fryone opened the pull request:
### Change target Owner on impact

- Warheads can now change targets owner to warhead's owner.
- `ChangeOwner.SetAsMindControl` makes the effect work like permanent mind control, which respects `ImmuneToPsionics`.
  - `ChangeOwner.MindControlAnim` determines the mind control anim of this effect, which respects `MindControlRingOffset`.

In `rulesmd.ini`:
```ini
[SOMEWARHEAD]                               ; WarheadType
ChangeOwner=false                           ; boolean
ChangeOwner.SetAsMindControl=false          ; boolean
ChangeOwner.MindControlAnim=                ; Animation
```

Fryone committed:
initial commit

added ownerchanger to warheads

Fryone committed:
Immunities (wip)

Immunities update

Fryone committed:
Docs and fixes

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge branch 'develop' into fryone/owner_changer

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/owner_changer

Fryone committed:
Update YRpp

brsajo commented:
The `IsOwnerChanger.ControlAnim` anim doesn't follow the unit around like the normal mindcontrol animation as I would have expected. If not a bug, could this be added as an option?

![image](https://github.com/Phobos-developers/Phobos/assets/86821333/dc0ed352-d5e6-44ae-b21b-8c1c7761d03c)

Fryone commented:
> The `IsOwnerChanger.ControlAnim` anim doesn't follow the unit around like the normal mindcontrol animation as I would have expected. If not a bug, could this be added as an option?
> 
weird, can you please provide settings for owner changer? As I tested, my anim is following techno as should

Fryone committed:
Update Detonate.cpp

chaserli commented:
It's a bit messed up.
If this is supposed to be a generic house change, fine, but do not mess up with existing MindControl or permanent mind control logic.
Even if it's for a generic house change, a simple ownership change vtbl call is not sufficient. I would suggest you look at how it's done for permanent mind control in YR and Ares in the first place.
Same issue with all your other pr.
btw it seems the bot is triggered upon every commit, so plz don't just merge from upstream constantly

Fryone committed:
fix permamc anim not following target

Fryone commented:
I'm not touching MC, instead I'm using PermaMC as it's much simpler. The only issue is that ownerchanged units without PermaMC set, allowed to enter in transports, makes their attached animation (if its looping) dissapear on leaving transport. So I'll just drop it for ares' attached anim instead.

Fryone committed:
updated logic

'IsOwnerChanger.ControlAnim' -> changed to 'IsOwnerChanger.MindControlAnim' to use specifically as custom anim for permanent MC

brsajo commented:
> weird, can you please provide settings for owner changer? As I tested, my anim is following techno as should

The nightly build seems to be failing so I can't test your latest changes but here's what I have
```
[Controller]
AnimList=YURICNTL
; MindControl=yes
Verses=100%,100%,100%,100%,100%,100%,0%,0%,0%,100%,100%
IsOwnerChanger=yes
IsOwnerChanger.SetAsMindControl=yes
IsOwnerChanger.ControlAnim=MINDANIM
IsOwnerChanger.AffectElites=yes
IsOwnerChanger.HealthThreshold=1.0
```
```
[MINDANIM]
LoopCount=-1
Rate=300
```
I also tested with EMP_FX01 and MINDANIMR (unchanged from vanilla).

Fryone commented:
Yep, as I thought, now 'ControlAnim' changed to 'MindControlAnim' and will follow target as it should.

Fryone committed:
Merge remote-tracking branch 'upstream/develop' into fryone/owner_changer

TaranDahl commented on src/Ext/WarheadType/Detonate.cpp:
```diff
@@ -250,6 +253,53 @@ void WarheadTypeExt::ExtData::ApplyRemoveDisguiseToInf(HouseClass* pHouse, Techn
+
+	if (!isImmune)
+	{
+		if ((GeneralUtils::GetWarheadVersusArmor(this->OwnerObject(), armorType) != 0.0) && (!pTarget->IsMindControlled()))
```
No need to check for versus, since `CanAffectTarget` will check it.

TaranDahl commented on src/Ext/WarheadType/Body.cpp:
```diff
@@ -127,6 +127,11 @@ void WarheadTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
+	this->ChangeOwner.Read(exINI, pSection, "IsOwnerChanger");
+	this->ChangeOwner_MindControl.Read(exINI, pSection, "IsOwnerChanger.SetAsMindControl");
+	this->ChangeOwner_MindAnim.Read(exINI, pSection, "IsOwnerChanger.MindControlAnim");
+	this->ChangeOwner_Threshold.Read(exINI, pSection, "IsOwnerChanger.HealthThreshold");
```
`AffectsBelowPercent` and `AffectsAbovePercent` have already done that.

TaranDahl reviewed (changes requested):
Fix the following things, then this can be merged.

Coronia committed:
Merge branch 'develop' into fryone/owner_changer

# Conflicts:
#	CREDITS.md
#	YRpp
#	docs/New-or-Enhanced-Logics.md
#	docs/Whats-New.md
#	src/Ext/TechnoType/Body.h
#	src/Ext/WarheadType/Body.cpp
#	src/Ext/WarheadType/Body.h
#	src/Ext/WarheadType/Detonate.cpp

Coronia committed:
fix mind control anim

Coronia commented:
changed tag names to `ChangeOwner` and removed repetitive versus and health check

I'm not sure about whether the elite check should be removed tho. There're only a few warhead effects which really need it, so maybe it don't have to be a general check

Fryone commented:
maybe `ChangeOwner.MindControlAnim` should be renamed to `ChangeOwner.ControlAnim` too

Coronia committed:
remove ChangeOwner.AffectElites

Coronia committed:
Merge remote-tracking branch 'origin/develop' into fryone/owner_changer

Coronia committed:
remove ChangeOwnerImmunities

Coronia reviewed (approved):
removed ChangeOwner.AffectElites due to the introduction of generic veterancy filter: https://github.com/Phobos-developers/Phobos/pull/2008

Fryone commented:
Is it worth to add timed control, similar to Dune's Deviator?

Coronia commented:
I think it could be useful, but we should also think about how to deal with the interaction of multiple ChangeOwner warheads and other owner changer effects (mind control, capture etc). Maybe it's an easier way to just make the timer be interrupted by another change house event (there's existing hook for that too)

Fryone committed:
Merge branch 'fryone/owner_changer' of https://github.com/Fryone/Phobos into fryone/owner_changer

Fryone committed:
Added timed control

- applied timed control if ChangeOwner.Duration is set
- no faulty checks for now

Coronia committed:
update ChangeOwner.Duration

Coronia committed:
Merge remote-tracking branch 'upstream/develop' into fryone/owner_changer

# Conflicts:
#	docs/Whats-New.md

Fryone commented:
I thought of something like levels, basic integer that orders different changers (change if new one is higher)

Coronia committed:
Merge branch 'develop' into fryone/owner_changer

# Conflicts:
#	docs/Whats-New.md
#	src/Ext/Techno/Body.Update.cpp
#	src/Ext/Techno/Body.cpp
#	src/Ext/Techno/Body.h
#	src/Ext/WarheadType/Body.cpp
#	src/Ext/WarheadType/Body.h
#	src/Ext/WarheadType/Detonate.cpp

Coronia committed:
remove ChangeOwner.Duration for now

Coronia committed:
update doc

Coronia committed:
Merge branch 'develop' into fryone/owner_changer
